ocaml/libs/tracing/tracing_export.ml

```diff
@@ -0,0 +1,339 @@
+          match !file_name with None -> make_file_name () | Some x -> x
+        in
+        Xapi_stdext_unix.Unixext.mkdir_rec (Filename.dirname file_name) 0o700 ;
+        let@ fd = file_name |> with_fd in
+        write fd json ;
+        if (Unix.fstat fd).st_size >= !max_file_size then (
+          debug "Tracing: Rotating file %s > %d" file_name !max_file_size ;
+          if !compress_tracing_files then
+            Zstd.Fast.compress_file Zstd.Fast.compress ~file_path:file_name
+              ~file_ext:"zst" ;
+          ignore @@ make_file_name ()
+        ) ;
+        Ok ()
+      with e -> Error e
+
+    let with_stream f =
+      Xapi_stdext_threads.Threadext.Mutex.execute lock (fun () -> f export)
+  end
+
+  module Http = struct
+    module Request = Cohttp.Request.Make (Cohttp_posix_io.Buffered_IO)
+    module Response = Cohttp.Response.Make (Cohttp_posix_io.Buffered_IO)
+
+    let export ~url json =
+      try
+        let body = json in
+        let content_headers =
+          [
+            ("Content-Type", "application/json")
+          ; ("Content-Length", string_of_int (String.length body))
+          ]
+        in
+        let host =
+          match (Uri.host url, Uri.port url) with
+          | None, _ ->
+              None
+          | Some host, None ->
+              Some host
+          | Some host, Some port ->
+              Some (Printf.sprintf "%s:%d" host port)
+        in
+        let host_headers =
+          Option.fold ~none:[] ~some:(fun h -> [("Host", h)]) host
+        in
+        let headers =
+          List.concat [content_headers; host_headers] |> Cohttp.Header.of_list
+        in
+
+        Open_uri.with_open_uri url (fun fd ->
+            let request =
+              Cohttp.Request.make ~meth:`POST ~version:`HTTP_1_1 ~headers url
+            in
+            (* `with_open_uri` already closes the `fd`. And therefore
+               according to the documentation of `in_channel_of_descr` and
+               `out_channel_of_descr` we should not close the channels on top of
+               `fd`. *)
+            let ic = Unix.in_channel_of_descr fd in
+            let oc = Unix.out_channel_of_descr fd in
+            Request.write
+              (fun writer -> Request.write_body writer body)
+              request oc ;
+            (* We flush instead of closing the sending stream as nginx responds to a TCP
+               half-shutdown with a full shutdown of both directions of the HTTP request *)
+            flush oc ;
+            match try Response.read ic with _ -> `Eof with
+            | `Eof ->
+                Ok ()
+            | `Invalid x ->
+                Error (Failure ("invalid read: " ^ x))
+            | `Ok response
+              when Cohttp.Code.(response.status |> code_of_status |> is_error)
+              ->
+                Error (Failure (Cohttp.Code.string_of_status response.status))
+            | `Ok response ->
+                let body = Buffer.create 128 in
```

Comment:
Where is the buffer content used?

Reply:
Previously the buffer content was included in the Ok result, but it was removed in this commit: https://github.com/xapi-project/xen-api/commit/daac640b5e782181766c2bdd22acd1e868d349cc#diff-fd80e1d35d536590437bc43e05e28fb1210e23db37ae6ae8c9ccf6e5e369d6b4L636 There isn't an explanation as to why, but maybe just because the Result message wasn't being used, but then why read the Response at all?

Reply:
https://github.com/xapi-project/xen-api/pull/5551/commits/bcc563a7cca7fd2a7ea8ac8194efa7b32c603057 I removed the read as it is not being used.